skrifa/src/color/mod.rs

```diff
@@ -0,0 +1,127 @@
+
+use instance::{resolve_paint, PaintId};
+
+/// Interface for receiving a sequence of commands that represent a flattened
+/// paint graph in a color outline.
+// Placeholder
+pub trait ColorPainter {}
+
+/// Affine transformation matrix.
+#[derive(Copy, Clone, Debug)]
+pub struct Transform {
+    xx: f32,
+    yx: f32,
+    xy: f32,
+    yy: f32,
+    dx: f32,
+    dy: f32,
+}
+
+/// Reference to a paint graph that represents a color glyph outline.
+#[derive(Clone)]
+pub struct ColorOutline<'a> {
+    colr: colr::Colr<'a>,
+    glyph_id: GlyphId,
+    kind: ColorOutlineKind<'a>,
+}
+
+impl<'a> ColorOutline<'a> {
+    /// Returns the glyph identifier that was used to retrieve this outline.
+    pub fn glyph_id(&self) -> GlyphId {
+        self.glyph_id
+    }
+
+    /// Evaluates the paint graph at the specified location in variation space
+    /// and returns the bounding box for the full scene.
+    ///
+    /// The `glyph_bounds` closure will be invoked for each clip node in the
+    /// graph and should return the bounding box for the given glyph, location
+    /// and transform.
+    pub fn bounding_box(
+        &self,
+        location: impl Into<LocationRef<'a>>,
+        glyph_bounds: impl FnMut(GlyphId, &LocationRef, Transform) -> Option<BoundingBox>,
+    ) -> Option<BoundingBox> {
+        None
+    }
+
+    /// Evaluates the paint graph at the specified location in variation space
+    /// and emits the results to the given painter.
+    pub fn paint(
+        &self,
+        location: impl Into<LocationRef<'a>>,
+        painter: &mut impl ColorPainter,
+    ) -> Result<(), ReadError> {
+        let instance = instance::ColrInstance::new(self.colr.clone(), location.into().coords());
+        match &self.kind {
+            ColorOutlineKind::V0(layer_range) => {
+                for layer_ix in layer_range.clone() {
+                    let (glyph_id, palette_ix) = instance.v0_layer(layer_ix)?;
+                }
+            }
+            ColorOutlineKind::V1(paint, paint_id) => {
+                let paint = resolve_paint(&instance, paint)?;
+            }
+        }
+        Ok(())
+    }
+}
+
+#[derive(Clone)]
+enum ColorOutlineKind<'a> {
+    V0(Range<usize>),
+    V1(colr::Paint<'a>, PaintId),
+}
+
+/// Collection of color outlines.
+#[derive(Clone)]
+pub struct ColorOutlineCollection<'a> {
+    glyph_count: u16,
+    colr: Option<colr::Colr<'a>>,
+}
+
+impl<'a> ColorOutlineCollection<'a> {
+    /// Creates a new color outline collection for the given font.
+    pub fn new(font: &impl TableProvider<'a>) -> Self {
+        let glyph_count = font
+            .maxp()
+            .map(|maxp| maxp.num_glyphs())
+            .unwrap_or_default();
+        let colr = font.colr().ok();
+        Self { glyph_count, colr }
+    }
+
+    /// Returns the color outline for the given glyph identifier.
+    pub fn get(&self, glyph_id: GlyphId) -> Option<ColorOutline<'a>> {
```

Comment:
I think the direct access is more important than the iterator - so perhaps we can do without the `maxp()` access? 

As discussed, in https://github.com/googlefonts/oxidize/pull/48 - we need the (cheaply acquirable) information in Skia whether a glyph is COLRv0 and COLRv1. I think that's doable well with the type-based distinction of `ColorOutlineKind`.

Reply:
I’d like to keep the iterator but I can implement it without maxp. Good catch.

Reply:
Re detection: I can add a `version()` method to `ColorOutline` that returns 0 or 1